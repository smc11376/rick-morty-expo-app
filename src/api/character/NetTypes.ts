export type CharacterNetType = {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: OriginOrLocationNetType;
  location: OriginOrLocationNetType;
  image: string;
  episode?: string[] | null;
  url: string;
  created: string;
};

export type OriginOrLocationNetType = {
  name: string;
  url: string;
};

export type GetAllCharactersNetType = {
  info: {
    count: number;
    pages: number;
    next: string | null;
    prev: string | null;
  };
  results: CharacterNetType[];
};
