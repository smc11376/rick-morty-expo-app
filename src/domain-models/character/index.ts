export type ItemWithUrl = {
  name: string;
  url: string;
};

export type ShortCharacter = {
  id: number;
  name: string;
  status: string;
  image: string;
};

export type Character = ShortCharacter & {
  species: string;
  type: string;
  gender: string;
  origin: ItemWithUrl;
  location: ItemWithUrl;
  numberOfEpisodes: number;
};
