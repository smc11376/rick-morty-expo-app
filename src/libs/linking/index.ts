import * as Linking from "expo-linking";

export function openUrl(url?: string) {
  if (!url) return;
  Linking.canOpenURL(url).then(canOpen => {
    if (canOpen) Linking.openURL(url);
  });
}
