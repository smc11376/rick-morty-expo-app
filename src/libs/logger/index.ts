export default {
  log: (message: any, ...optionalParams: any[]) => {
    if (__DEV__) {
      console.log(new Date().toLocaleString(), message, ...optionalParams);
    }
  },
  info: (message: any, ...optionalParams: any[]) => {
    if (__DEV__) {
      console.info(new Date().toLocaleString(), message, ...optionalParams);
    }
  },
  error: (message: any, ...optionalParams: any[]) => {
    if (__DEV__) {
      console.error(new Date().toLocaleString(), message, ...optionalParams);
    }
  },
  warn: (message: any, ...optionalParams: any[]) => {
    if (__DEV__) {
      console.warn(new Date().toLocaleString(), message, ...optionalParams);
    }
  },
};
