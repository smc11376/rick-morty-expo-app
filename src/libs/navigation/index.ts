import { router } from "expo-router";
import { CHARACTER_DETAIL_PATH, HOME_PATH } from "@/libs/navigation/pathRoutes";

function goToPath(path: string, params?: any) {
  router.navigate(path);
  router.setParams(params);
}

export function goBack() {
  router.back();
}

export function goToHome() {
  goToPath(HOME_PATH);
}

export function goToCharacterDetail(id: number, characterName: string) {
  goToPath(CHARACTER_DETAIL_PATH.replace("[id]", id.toString()), { characterName });
}
