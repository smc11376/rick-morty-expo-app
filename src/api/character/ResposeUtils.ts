import { CharacterNetType, GetAllCharactersNetType } from "@/api/character/NetTypes";
import { Character, ShortCharacter } from "@/domain-models/character";

export function transformGetAllCharactersResponse(response?: GetAllCharactersNetType): ShortCharacter[] {
  if (!response) return [];

  return response.results.map(item => ({
    id: item.id,
    name: item.name,
    status: item.status,
    image: item.image,
  }));
}

export function transformGetCharacterByIdResponse(response?: CharacterNetType): Character | undefined {
  if (!response) return undefined;

  return {
    id: response.id,
    name: response.name,
    status: response.status,
    species: response.species,
    type: response.type,
    gender: response.gender,
    origin: response.origin,
    location: response.location,
    image: response.image,
    numberOfEpisodes: response.episode?.length || 0,
  };
}
