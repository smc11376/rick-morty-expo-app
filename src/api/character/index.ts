import { useGetApiService } from "@/libs/network";
import { CharacterNetType, GetAllCharactersNetType } from "@/api/character/NetTypes";
import { transformGetAllCharactersResponse, transformGetCharacterByIdResponse } from "@/api/character/ResposeUtils";
import { Character, ShortCharacter } from "@/domain-models/character";
import { useEffect, useMemo, useState } from "react";

type UseAllCharactersType = {
  characters: ShortCharacter[];
  nextPage: string | null | undefined;
  isLoading: boolean;
};

type UseCharacterType = {
  character: Character | undefined;
  isLoading: boolean;
};

export function useGetAllCharacters(newUrl: string | null | undefined): UseAllCharactersType {
  const [characters, setCharacters] = useState<ShortCharacter[]>([]);
  const { data, isLoading, isFetching, isRefetching } = useGetApiService<GetAllCharactersNetType>(
    newUrl || "/character",
  );
  const newCharacters = useMemo(() => transformGetAllCharactersResponse(data), [data?.results]);

  useEffect(() => {
    if (!isLoading && !isFetching && !isRefetching) {
      setCharacters(currentCharacters => {
        return [...new Set([...currentCharacters, ...newCharacters])];
      });
    }
  }, [newCharacters]);

  return {
    characters,
    isLoading: isLoading || isFetching || isRefetching,
    nextPage: data?.info.next,
  };
}

export function useGetCharacterById(id?: string): UseCharacterType {
  if (!id) throw new Error("Oops! There is no character to find!");

  const { data, isLoading, isFetching, isRefetching } = useGetApiService<CharacterNetType>(`/character/${id}`);
  const character = transformGetCharacterByIdResponse(data);
  return {
    character,
    isLoading: isLoading || isFetching || isRefetching,
  };
}
