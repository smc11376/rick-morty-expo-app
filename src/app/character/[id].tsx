import { ScrollView, StyleSheet, View } from "react-native";
import { useGlobalSearchParams, useNavigation } from "expo-router";
import { useGetCharacterById } from "@/api/character";
import { useEffect } from "react";
import Loader from "@/components/loader";
import Row from "@/app/character/components/row";
import InfoBlock from "@/app/character/components/info-block";
import CharacterImage from "@/app/character/components/character-image";

export default function CharacterDetailScreen() {
  const params = useGlobalSearchParams();
  const { character, isLoading } = useGetCharacterById(params.id as string);
  const navigation = useNavigation();

  useEffect(() => {
    navigation.setOptions({ title: params.characterName });
  }, [params.characterName]);

  return isLoading ? (
    <View style={styles.container}>
      <Loader />
    </View>
  ) : (
    <ScrollView style={styles.container} contentContainerStyle={styles.contentContainerStyle}>
      <CharacterImage uri={character?.image} />
      <Row>
        <InfoBlock label="Gender: " data={character?.gender} />
        <InfoBlock label="Status: " data={character?.status} />
      </Row>
      <Row>
        <InfoBlock label="Species: " data={character?.species} />
        {character?.type !== "" && <InfoBlock label="Type: " data={character?.type} />}
      </Row>
      <InfoBlock label="Number of episodes: " data={character?.numberOfEpisodes.toString()} />
      <InfoBlock label="Origin: " data={character?.origin.name} urlToOpen={character?.origin.url} />
      <InfoBlock label="Current location: " data={character?.location.name} urlToOpen={character?.location.url} />
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "rgb(60, 62, 68)",
  },
  contentContainerStyle: {
    flexGrow: 1,
  },
  image: {
    width: "100%",
    aspectRatio: 1,
  },
});
