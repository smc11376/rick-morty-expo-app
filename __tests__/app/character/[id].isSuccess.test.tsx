import { fireEvent, render, screen, waitFor } from "@testing-library/react-native";
import CharacterDetailScreen from "@/app/character/[id]";
import { Character } from "@/domain-models/character";
import * as Linking from "@/libs/linking";

const characterMock: Character = {
  id: 1,
  name: "Testing Name",
  status: "Alive",
  image: "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
  species: "Testing species",
  type: "Testing types",
  gender: "Testing gender",
  origin: {
    name: "Testing name",
    url: "Testing url",
  },
  location: {
    name: "Testing name",
    url: "Testing url",
  },
  numberOfEpisodes: 22,
};

jest.mock("@/api/character", () => ({
  useGetCharacterById: () => ({
    character: characterMock,
    isLoading: false,
  }),
}));

test("CharacterDetailScreen renders isSuccess", () => {
  render(<CharacterDetailScreen />);

  expect(screen.toJSON()).toMatchSnapshot();
});

test("CharacterDetailScreen actions", async () => {
  const openUrlMock = jest.spyOn(Linking, "openUrl");
  render(<CharacterDetailScreen />);

  fireEvent.press(screen.getAllByTestId("info-block-url")[0]);
  await waitFor(() => expect(openUrlMock).toHaveBeenCalled());
});
