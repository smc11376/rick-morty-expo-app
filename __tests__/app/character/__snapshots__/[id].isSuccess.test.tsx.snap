// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetailScreen renders isSuccess 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "flexGrow": 1,
    }
  }
  style={
    {
      "backgroundColor": "rgb(60, 62, 68)",
      "flex": 1,
    }
  }
>
  <View>
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
        }
      }
      style={
        {
          "aspectRatio": 1,
          "width": "100%",
        }
      }
    />
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          [
            {
              "padding": 16,
              "paddingBottom": 0,
            },
            {
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          style={
            {
              "color": "grey",
              "fontSize": 20,
            }
          }
        >
          Gender: 
        </Text>
        <Text
          onPress={[Function]}
          style={
            {
              "color": "white",
              "fontSize": 20,
            }
          }
          testID="info-block-url"
        >
          Testing gender
        </Text>
      </View>
      <View
        style={
          [
            {
              "padding": 16,
              "paddingBottom": 0,
            },
            {
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          style={
            {
              "color": "grey",
              "fontSize": 20,
            }
          }
        >
          Status: 
        </Text>
        <Text
          onPress={[Function]}
          style={
            {
              "color": "white",
              "fontSize": 20,
            }
          }
          testID="info-block-url"
        >
          Alive
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          [
            {
              "padding": 16,
              "paddingBottom": 0,
            },
            {
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          style={
            {
              "color": "grey",
              "fontSize": 20,
            }
          }
        >
          Species: 
        </Text>
        <Text
          onPress={[Function]}
          style={
            {
              "color": "white",
              "fontSize": 20,
            }
          }
          testID="info-block-url"
        >
          Testing species
        </Text>
      </View>
      <View
        style={
          [
            {
              "padding": 16,
              "paddingBottom": 0,
            },
            {
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          style={
            {
              "color": "grey",
              "fontSize": 20,
            }
          }
        >
          Type: 
        </Text>
        <Text
          onPress={[Function]}
          style={
            {
              "color": "white",
              "fontSize": 20,
            }
          }
          testID="info-block-url"
        >
          Testing types
        </Text>
      </View>
    </View>
    <View
      style={
        [
          {
            "padding": 16,
            "paddingBottom": 0,
          },
          {
            "flexDirection": "row",
          },
        ]
      }
    >
      <Text
        style={
          {
            "color": "grey",
            "fontSize": 20,
          }
        }
      >
        Number of episodes: 
      </Text>
      <Text
        onPress={[Function]}
        style={
          {
            "color": "white",
            "fontSize": 20,
          }
        }
        testID="info-block-url"
      >
        22
      </Text>
    </View>
    <View
      style={
        [
          {
            "padding": 16,
            "paddingBottom": 0,
          },
          false,
        ]
      }
    >
      <Text
        style={
          {
            "color": "grey",
            "fontSize": 20,
          }
        }
      >
        Origin: 
      </Text>
      <Text
        onPress={[Function]}
        style={
          {
            "color": "white",
            "fontSize": 20,
          }
        }
        testID="info-block-url"
      >
        Testing name
      </Text>
    </View>
    <View
      style={
        [
          {
            "padding": 16,
            "paddingBottom": 0,
          },
          false,
        ]
      }
    >
      <Text
        style={
          {
            "color": "grey",
            "fontSize": 20,
          }
        }
      >
        Current location: 
      </Text>
      <Text
        onPress={[Function]}
        style={
          {
            "color": "white",
            "fontSize": 20,
          }
        }
        testID="info-block-url"
      >
        Testing name
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
